File: framework/ndn_iot_cpp/examples/test-consuming.cpp
#include "bootstrap.hpp"
#include "common.hpp"
#include "boost/asio.hpp"

using namespace ndn;
using namespace std;
using namespace ndn::func_lib;

namespace ndn_iot {

namespace examples {

class AppConsumer
{
public:
  AppConsumer() {
    face_ = ptr_lib::shared_ptr<ThreadsafeFace>(new ThreadsafeFace(ioService_));
  }

  void run() {
    cout << "Running test consumer" << endl;
    Bootstrap bootstrap(*(face_.get()));
    keyChain_ = bootstrap.setupDefaultIdentityAndRoot(Name("/home/mygooddevice"));
    Name defaultCertificate = bootstrap.getDefaultCertificateName();
    bootstrap.startTrustSchemaUpdate(Name("/home/gateway/flow"), 
      bind(&AppConsumer::onUpdateSuccess, this, _1, _2), 
      bind(&AppConsumer::onUpdateFailed, this, _1));

    boost::asio::io_service::work work(ioService_);
    ioService_.run();
  }
private:
  void onUpdateSuccess(string schema, bool isInitial) {
    cout << "trust schema update success" << endl;
    if (isInitial) {
      face_->expressInterest(Name("/home/flow/cpp-publisher-1"), 
        bind(&AppConsumer::onData, this, _1, _2),
        bind(&AppConsumer::onTimeout, this, _1));
    }
  }

  void onUpdateFailed(string msg) {
    cout << "Trust schema update failed: " << msg << endl;
  }

  void onVerified(const ptr_lib::shared_ptr<const Data>& data) {
    cout << "data verified: " << data->getName().toUri() << endl;
  }

  void onVerifyFailed(const ptr_lib::shared_ptr<const Data>& data) {
    cout << "data verification failed: " << data->getName().toUri() << endl;
  }

  void onData(const ptr_lib::shared_ptr<const Interest>& interest, const ptr_lib::shared_ptr<Data>& data) {
    keyChain_->verifyData(data, 
      bind(&AppConsumer::onVerified, this, _1),
      (const OnVerifyFailed)bind(&AppConsumer::onVerifyFailed, this, _1));
  }

  void onTimeout(const ptr_lib::shared_ptr<const Interest>& interest) {
    cout << "interest times out: " << interest->getName().toUri() << endl;
  }

  ptr_lib::shared_ptr<ThreadsafeFace> face_;
  boost::asio::io_service ioService_;
  ptr_lib::shared_ptr<KeyChain> keyChain_;
};

}

}

int main() {
  ndn_iot::examples::AppConsumer consumer;
  consumer.run();

  return 0;
}

File: framework/ndn_iot_dot_net/examples/test-consuming.cs
namespace ndn_iot.tests {
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using System.IO;

    // TODO: to be finished up with consumer functions as in other languages
    using System.Runtime.InteropServices;

    using net.named_data.jndn.security.policy;    
    using net.named_data.jndn;
    using net.named_data.jndn.encoding;
    using net.named_data.jndn.util;
    using net.named_data.jndn.security;
    using net.named_data.jndn.security.identity;
    using net.named_data.jndn.security.certificate;
    using net.named_data.jndn.encoding.tlv;

    using net.named_data.jndn.transport;

    using ndn_iot.bootstrap;

    class AppConsumer {
        class VerificationHandler : OnVerified, OnDataValidationFailed {
            public void onVerified(Data data) {
                Console.Out.WriteLine("Data verified");
            }

            public void onDataValidationFailed(Data data, string reason) {
                Console.Out.WriteLine("Data verification failed: " + reason);
            }
        }

        class DataHandler : OnData, OnTimeout {
            public DataHandler(AppConsumer ac) {
                ac_ = ac;
            }

            public void onData(Interest interest, Data data) {
                Console.Out.WriteLine("data received: " + data.getName().toUri() + "; verifying!");
                VerificationHandler vh = new VerificationHandler();
                ac_.keyChain_.verifyData(data, vh, vh);
            }

            public void onTimeout(Interest interest) {
                Console.Out.WriteLine("interest timed out: " + interest.getName().toUri());
            }

            AppConsumer ac_;
        }

        void onUpdateSuccess(string schema, bool isInitial) {
            Console.Out.WriteLine("trust schema update success");
            DataHandler dh = new DataHandler(this);

            if (isInitial) {
                face_.expressInterest(new Name("/home/flow/cpp-publisher-1"), dh, dh);
            }
        }

        void onUpdateFailed(string msg) {
            Console.Out.WriteLine("Trust schema update failed: " + msg);
        }

        public void run()
        {
            face_ = new Face(new TcpTransport(), new TcpTransport.ConnectionInfo("localhost"));
            Bootstrap bootstrap = new Bootstrap(face_);
            keyChain_ = bootstrap.setupDefaultIdentityAndRoot(new Name("/home/flow-csharp"), new Name());

            bootstrap.startTrustSchemaUpdate(new Name("/home/gateway/flow"), onUpdateSuccess, onUpdateFailed);
            while (true) {
                face_.processEvents();
                // We need to sleep for a few milliseconds so we don't use 100% of the CPU.
                System.Threading.Thread.Sleep(5);
            }
        }

        KeyChain keyChain_;
        Face face_;
    }

    class TestConsumer {
        static void Main(string[] args) {
            AppConsumer apConsumer = new AppConsumer();
            apConsumer.run();
        }
    }
}